api/proto/auth/contextdialer.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"net"
+	"time"
+)
+
+// ContextDialer represents network dialer interface that uses context
+type ContextDialer interface {
+	// DialContext is a function that dials to the specified address
+	DialContext(ctx context.Context, network, addr string) (net.Conn, error)
+}
+
+// ContextDialerFunc is a function wrapper that implementsa
+// ContextDialer interface
+type ContextDialerFunc func(ctx context.Context, network, addr string) (net.Conn, error)
+
+// DialContext is a function that dials to the specified address
+func (f ContextDialerFunc) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	return f(ctx, network, addr)
+}
+
+// NewAddrDialer returns new dialer from a list of addresses
+func NewAddrDialer(addrs []string, keepAliveInterval time.Duration, dialTimeout time.Duration) ContextDialer {
+	if len(addrs) == 0 {
+		panic("no addresses to dial")
```

Comment:
Why not return an error instead?
Alternatively, call this `MustNewAddrDialer` to make it clear that this panics.